File: backend/lib/vectorstore/resumestore.py
from pathlib import Path
from langchain.document_loaders import PyPDFLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_core.vectorstores import InMemoryVectorStore
from langchain_openai import OpenAIEmbeddings

_resume_vector_store = None

def get_resume_store():
    """
    Load and cache the resume PDF as an in-memory vector store.
    """
    global _resume_vector_store
    if _resume_vector_store is not None:
        print("✅ [Resume Store] Using cached resume vector store")
        return _resume_vector_store

    print("📄 [Resume Store] Loading and embedding resume PDF...")

    # PDF expected at backend/assets/resume.pdf
    pdf_path = Path(__file__).parent.parent.parent / "assets" / "resume.pdf"
    if not pdf_path.exists():
        raise FileNotFoundError(f"Resume PDF not found at {pdf_path}")

    loader = PyPDFLoader(str(pdf_path))
    docs = loader.load()
    splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
    chunks = splitter.split_documents(docs)

    embeddings = OpenAIEmbeddings()
    store = InMemoryVectorStore.from_documents(chunks, embeddings)

    _resume_vector_store = store
    print("🧠 [Resume Store] Resume embedded and cached successfully")
    return store

File: frontend/lib/vectorstore/resumestore.ts
import { PDFLoader } from "@langchain/community/document_loaders/fs/pdf";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import { MemoryVectorStore } from "langchain/vectorstores/memory";
import { OpenAIEmbeddings } from "@langchain/openai";

const globalForVectorStore = globalThis as unknown as {
  resumeVectorStore?: MemoryVectorStore;
};

export async function getResumeStore() {
  if (globalForVectorStore.resumeVectorStore) {
    console.log("✅ [Resume Store] Using cached resume vector store");
    return globalForVectorStore.resumeVectorStore;
  }

  console.log("📄 [Resume Store] Loading and embedding resume PDF...");

  const loader = new PDFLoader("public/assets/resume.pdf");
  const docs = await loader.load();

  const splitter = new RecursiveCharacterTextSplitter({
    chunkSize: 1000,
    chunkOverlap: 200,
  });

  const chunks = await splitter.splitDocuments(docs);
  const embeddings = new OpenAIEmbeddings({ apiKey: process.env.OPENAI_API_KEY });

  const store = await MemoryVectorStore.fromDocuments(chunks, embeddings);
  globalForVectorStore.resumeVectorStore = store;

  console.log("🧠 [Resume Store] Resume embedded and cached successfully");

  return store;
}
